File: typescript/packages/core/src/context-compactor.ts
import type { ContextCompactor } from './types/contracts.js';
import type { Message } from './types/domain.js';

function estimateTokens(message: Message): number {
  const text =
    typeof message.content === 'string'
      ? message.content
      : JSON.stringify(message.content);
  return Math.ceil(text.length / 4);
}

export class TruncateMiddleCompactor implements ContextCompactor {
  async compact(messages: Message[], tokenBudget: number): Promise<Message[]> {
    if (messages.length <= 2) {
      return messages;
    }

    const system = messages.filter((message) => message.role === 'system');
    const rest = messages.filter((message) => message.role !== 'system');

    let total = [...system, ...rest].reduce((sum, message) => sum + estimateTokens(message), 0);
    if (total <= tokenBudget) {
      return messages;
    }

    const head = rest.slice(0, 1);
    const tail = rest.slice(-4);
    const dropped = rest.length - head.length - tail.length;

    const summary: Message = {
      role: 'system',
      content: `[Context compacted: ${dropped} middle messages summarized to stay within token budget.]`,
    };

    const compacted = [...system, ...head, summary, ...tail];
    total = compacted.reduce((sum, message) => sum + estimateTokens(message), 0);

    while (total > tokenBudget && compacted.length > 2) {
      const removable = compacted.findIndex(
        (message, index) => message.role !== 'system' && index > 0 && index < compacted.length - 2,
      );
      if (removable === -1) {
        break;
      }
      compacted.splice(removable, 1);
      total = compacted.reduce((sum, message) => sum + estimateTokens(message), 0);
    }

    return compacted;
  }
}

export { estimateTokens };

File: python/src/agent_framework/core/context_compactor.py
from __future__ import annotations

from agent_framework.core.domain import Message


def estimate_tokens(message: Message) -> int:
    content = message.content if isinstance(message.content, str) else str(message.content)
    return max(1, len(content) // 4)


class TruncateMiddleCompactor:
    async def compact(self, messages: list[Message], token_budget: int) -> list[Message]:
        if len(messages) <= 2:
            return messages

        system = [message for message in messages if message.role == "system"]
        rest = [message for message in messages if message.role != "system"]
        total = sum(estimate_tokens(message) for message in messages)
        if total <= token_budget:
            return messages

        head = rest[:1]
        tail = rest[-4:]
        dropped = len(rest) - len(head) - len(tail)
        summary = Message(
            role="system",
            content=f"[Context compacted: {dropped} middle messages summarized to stay within token budget.]",
        )
        compacted = [*system, *head, summary, *tail]
        while sum(estimate_tokens(message) for message in compacted) > token_budget and len(compacted) > 2:
            removable = next(
                (
                    index
                    for index, message in enumerate(compacted)
                    if message.role != "system" and 0 < index < len(compacted) - 2
                ),
                None,
            )
            if removable is None:
                break
            compacted.pop(removable)
        return compacted
